DOCSP-31150: Remove legacy projects from PHP libraries (#889) (#892)
mongodb/docs-ecosystem: source/php-libraries.txt

PHP Libraries, Frameworks, and Tools
====================================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Standalone Libraries
--------------------

- `Doctrine MongoDB ODM <https://github.com/doctrine/mongodb-odm>`__ (Object
  Document Mapper) is a library that provides object-mapping functionality for
  MongoDB.  You can use the standalone library or use one of the following
  framework integrations:

  - `Symfony <https://github.com/doctrine/DoctrineMongoDBBundle>`__
  - `Laminas <https://github.com/doctrine/DoctrineMongoODMModule>`__ (formerly Zend Framework)

- `Mongo PHP Adapter <https://github.com/alcaeus/mongo-php-adapter>`__ is a

>   userland library designed to act as an adapter between applications that rely

Reviewer: ```suggestion
  library designed to act as an adapter between applications that rely
```

   :depth: 2
   :class: singlecol

Standalone Libraries
--------------------

- `Doctrine MongoDB ODM <https://github.com/doctrine/mongodb-odm>`__ is a

>   library that provides object mapping functionality for MongoDB.

Reviewer 2: S: [compound modifiers](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#hyphens-in-compound-modifiers)

```suggestion
  library that provides object-mapping functionality for MongoDB.
```

>   You can use this library standalone or use one of the following framework

Reviewer 2: S: I know exactly what this means, but I'e never seen "standalone" used as an adverb

```suggestion
  You can use this library on its own or use one of the following framework
```

(or something similar)

Reviewer 2: Applies to all occurrences of "standalone" as adverb

  integrations:

  - `Symfony <https://github.com/doctrine/DoctrineMongoDBBundle>`__
  - `Laminas <https://github.com/doctrine/DoctrineMongoODMModule>`__ (formerly Zend Framework)

- `Mongo PHP Adapter <https://github.com/alcaeus/mongo-php-adapter>`__ is a

>   userland library designed to act as an adapter between applications that rely

Reviewer 2: Q: I was able to correctly guess what "userland" means but have never seen the term before. Do you think it will be clear enough for readers?

Author: I think it's tech and PHP jargon, which we should avoid if possible.

@GromNaN: Is this something that we can reword "userland library" to something like "is a library that does not have any dependencies"?

Reviewer: "Userland" is used throughout the PHP docs (e.g. [Userland Naming Guide](https://www.php.net/manual/en/userlandnaming.php)) and refers to any code written in PHP itself. It is distinct from PHP interpreter (language engine) and C extensions such as PHPC. [This Stack Overflow answer](https://stackoverflow.com/a/62336351/162228) elaborates in more detail.

It would be be incorrect to change "userland" to "a library that does not have any dependencies" but I don't object to just deleting it from this sentence if you'd rather it not appear in the docs. IMO the sentence reads just fine without it, as most readers should be able to discern that the linked package is a PHP library and not a C extension.

----

Note: the term also appears in [Legacy Driver Upgrade Guide](https://www.mongodb.com/docs/php-library/current/upgrade/#accessing-ids-of-inserted-documents) within the PHPLIB docs:

> it directly modified memory through the extension API and could not be implemented in PHP userland.

I'm not concerned about the usage there as I think its meaning is more easily inferred in the context of that sentence.

Author: Thanks for the explanation! From our style guide's guidance, we try to avoid jargon if there isn't an appropriate place to explain it or if it's not strictly necessary. I guess if we ever needed to explain the term, we would identify that it depends only on components within the PHP ecosystem such as built-ins (core?) and extensions?

For this instance, we'll take your suggestion of omitting it from the description.

Reviewer: > I guess if we ever needed to explain the term, we would identify that it depends only on components within the PHP ecosystem such as built-ins (core?) and extensions?

The dependencies, or lack thereof, are not relevant. "Userland" only refers to something implemented in PHP code (like [mongo-php-library](https://github.com/mongodb/mongo-php-library)), as opposed to being a C extension (like [mongo-php-driver](https://github.com/mongodb/mongo-php-driver)) or part of the PHP core language (also implemented in C). 

> since we don't want to provide a full explanation in a list of resources, we'll take your suggestion of omitting it from the description.

Sounds good.

  on the legacy ``mongo`` extension and the new ``mongodb`` extension. It
  offers the API of the legacy driver for the new driver and library.

- `Mongolid <https://github.com/leroy-merlin-br/mongolid>`__ is a performant
  ODM for PHP and MongoDB. It implements both ActiveRecord and DataMapper
  design patterns and supports embedded and referenced documents. You can use
  this library standalone or with the `Laravel <https://github.com/leroy-merlin-br/mongolid-laravel>`__
  integration.

- `Xenus <https://github.com/abellion/xenus>`__ is a MongoDB ODM
  that supports events, relationships, embedded documents, and more. You can
  use this library standalone or with the `Laravel <https://github.com/abellion/xenus-laravel>`__

>   integration which adds support for failed jobs, migrations, and events.

Reviewer 2: I: [Need comma before nonrestrictive clause](https://www.mongodb.com/docs/meta/style-guide/style/punctuation/#commas)

```suggestion
  integration, which adds support for failed jobs, migrations, and events.
```

Framework Integrations
----------------------

- Drupal

  - `MongoDB integration for Drupal <https://www.drupal.org/project/mongodb>`__.
    This is a collection of several modules which allow sites to store different

>     types of Drupal data in MongoDB. The ``mongodb`` extension supports

Reviewer 2: Out of scope but on the topic of restrictive clauses, this sentence should read, "This is a collection of several modules _that_ allow..."

    Drupal 8 and later.

- Laravel

  - `Laravel MongoDB <https://github.com/jenssegers/laravel-mongodb>`__ is an
    Eloquent model and Query builder that supports MongoDB by using the
    original Laravel API. This library extends the original Laravel classes
    and therefore uses the same methods.

- Symfony

  - You can configure the `Lock <https://symfony.com/doc/current/components/lock.html#mongodbstore>`__
    and the `Session <https://symfony.com/doc/current/session.html#store-sessions-in-a-nosql-database-mongodb>`__
    to use MongoDB as a data store.

  - `MongoDB Bundle <https://github.com/facile-it/mongodb-bundle>`__ is a
    bundle service integration for the official `PHP library <https://github.com/mongodb/mongo-php-library>`__.
    You can use it to configure connections to different databases or clusters.
    This integration includes a query profiler.

  - `DoctrineMongoDBBundle Symfony <https://github.com/doctrine/DoctrineMongoDBBundle>`__

>     This bundle integrates the Doctrine Object Document Mapper (ODM) into Symfony so

Reviewer 2: S: Put `(ODM)` at the first occurrence on this page (either line 18 or 32)

Author: Good suggestion. On this page the first usage is in a title in which the product name is "ODM", so I'll introduce it there accordingly.
I also noticed that this line needs to match the product name. I had taken it from the cherry-picked commit and didn't catch it.

    that you can store and retrieve objects from MongoDB.

- Yii2

  - `MongoDB Extension for Yii 2 <https://www.yiiframework.com/extension/yiisoft/yii2-mongodb>`__
    is a MongoDB integration for the Yii 2 framework.

Tools and Projects
------------------

- `PHP Cache <https://github.com/php-cache/mongodb-adapter/>`__ is a PSR-6 cache
  implementation that uses MongoDB as a cache pool. This project is part of
  the PHP Cache organization.

- `PHPfastcache <https://github.com/PHPSocialNetwork/phpfastcache>`__ is a
  high-performance backend cache system for MongoDB.

> - `Enqueue <https://github.com/php-enqueue/mongodb>`__ is a production ready

Reviewer 2: ```suggestion
- `Enqueue <https://github.com/php-enqueue/mongodb>`__ is a production-ready
```

  messaging solution that uses MongoDB as the message queue broker. It provides

>   a common way for programs to create, send, read messages.

Reviewer 2: I: missing conjunction

```suggestion
  a common way for programs to create, send, and read messages.
```

- `XHGui <https://github.com/perftools/xhgui>`__ is a web interface for the

>   XHProf profiler which stores profiling data in MongoDB.

Reviewer 2: I: Seems like this is a nonrestrictive clause, requiring a comma. (If it's restrictive, change to "that" and omit comma.)

```suggestion
  XHProf profiler, which stores profiling data in MongoDB.
```

Author: It provides additional detail to XMProf, so I think it is a nonrestrictive clause.
